homeassistant/components/roku/config_flow.py

```diff
@@ -0,0 +1,134 @@
+    """Handle a Roku config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = CONN_CLASS_LOCAL_POLL
+
+    @callback
+    def _show_form(self, errors: Optional[Dict] = None) -> Dict[str, Any]:
+        """Show the form to the user."""
+        return self.async_show_form(
+            step_id="user", data_schema=DATA_SCHEMA, errors=errors or {},
+        )
+
+    async def async_step_import(
+        self, user_input: Optional[Dict] = None
+    ) -> Dict[str, Any]:
+        """Handle configuration by yaml file."""
+        return await self.async_step_user(user_input)
+
+    async def async_step_user(
+        self, user_input: Optional[Dict] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initialized by the user."""
+        if not user_input:
+            return self._show_form()
+
+        errors = {}
+
+        try:
+            info = await self.hass.async_add_executor_job(validate_input, user_input)
+        except (SocketGIAError, RequestException, RokuException):
+            errors["base"] = ERROR_CANNOT_CONNECT
+            return self._show_form(errors)
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unknown error trying to connect.")
+            return self.async_abort(reason=ERROR_UNKNOWN)
+
+        await self.async_set_unique_id(info["serial_num"])
+        self._abort_if_unique_id_configured()
+
+        return self.async_create_entry(title=info["title"], data=user_input)
+
+    async def async_step_ssdp(
+        self, discovery_info: Optional[Dict] = None
+    ) -> Dict[str, Any]:
+        """Handle a flow initialized by discovery."""
+        host = urlparse(discovery_info[ATTR_SSDP_LOCATION]).hostname
+        name = discovery_info[ATTR_UPNP_FRIENDLY_NAME]
+        serial_num = discovery_info[ATTR_UPNP_SERIAL]
+
+        await self.async_set_unique_id(serial_num)
+        self._abort_if_unique_id_configured(updates={CONF_HOST: host})
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context.update(
+            {CONF_HOST: host, CONF_NAME: name, "title_placeholders": {"name": host}}
+        )
+
+        return await self.async_step_ssdp_confirm()
+
+    async def async_step_ssdp_confirm(
+        self, user_input: Optional[Dict] = None
+    ) -> Dict[str, Any]:
+        """Handle user-confirmation of discovered device."""
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        name = self.context.get(CONF_NAME)
+
+        if user_input is not None:
+            # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+            user_input[CONF_HOST] = self.context.get(CONF_HOST)
+            user_input[CONF_NAME] = name
+
+            try:
+                await self.hass.async_add_executor_job(validate_input, user_input)
+                return self.async_create_entry(title=name, data=user_input)
+            except (SocketGIAError, RequestException, RokuException):
```

Comment:
If we catch this inside `validate_input` instead and raise `CannotConnect` the code will be a bit more DRY.